### PR TITLE
[web-ui] Quick fixes for web-ui

### PR DESCRIPTION
**Pull request by yoannmoinet**
### What and why?

Related to web-ui's integration.
Upgrading it to latest versions.
Some small fixes where necessary.

### How?

- esbuild: stub file was being removed too soon and sub builds would fail on resolution.
  - instead of creating the file in the outDir, we now use a tmpdir, and we don't delete it.
- rspack: removal of stub file wasn't happening in case of failed build (using `shutdown` instead of `done` now).
  - using `shutdown` async would randomly crash too, so made it sync.
- rspack: re-organised the code to be all at the same place.
- `yarn cli prepare-link` wasn't reverting to normal after a slight drift from `package.json` latest definition.
- small simplification of `runBundlers` by removing the `isolateModules` usage.
- fix the build-report to account for externals better and report them.
- move a bunch of helpers in the core package.
- add tests to some missing helpers.
- fix the `env` var to correctly fallback to `development`.

### Follow-ups

Write a test that would catch the esbuild issue better.
The current test might be running too fast and doesn't reproduce the edge case well enough.

### CODE REVIEW

File: packages/core/src/helpers.ts [hunk 1]
```diff
@@ -212,11 +218,18 @@ export const truncateString = (
...
+// From a bundler's name, is it part of the "xpack" family?
+export const isXpack = (bundlerName: BundlerFullName) =>
+    ['rspack', 'webpack4', 'webpack5', 'webpack'].includes(bundlerName);
+
 // Replacing fs-extra with local helpers.
 // Delete folders recursively.
 export const rm = async (dir: string) => {
     return fsp.rm(dir, { force: true, maxRetries: 3, recursive: true });
 };
+export const rmSync = async (dir: string) => {
```
Comment (jakub-g) [inline packages/core/src/helpers.ts:230]:
```suggestion
export const rmSync = (dir: string) => {
```

Comment (jakub-g) [reply packages/core/src/helpers.ts:230]: (maybe some usages to update with `await` to be removed? though unlikely)

File: packages/factory/src/helpers.ts [hunk 2]
```diff
@@ -118,7 +118,9 @@ export const getContext = ({
         },
     };
 
-    const passedEnv: Env = (process.env.BUILD_PLUGINS_ENV as Env) || 'development';
+    // Use "production" if there is no env passed.
+    const passedEnv: Env = (process.env.BUILD_PLUGINS_ENV as Env) || 'production';
+    // Fallback to "development" if the passed env is wrong.
```
Comment (jakub-g) [inline packages/factory/src/helpers.ts:123]: Just curious, what's the difference between development and production here? Log verbosity? Something more?

Comment (yoannmoinet) [reply packages/factory/src/helpers.ts:123]: Right now, the main diff is logging to Datadog in the analytics plugin, only done in `production`.

About this change specifically: 
Before, it was always falling back to `development` when not defined.
And I forgot to explicitly define it when bundling and publishing it.
Resulting with a `development` env in actual production, so got me thinking that maybe the better default is `production`.
Jury's still out though.

File: packages/core/src/helpers.ts (hunk 1 above)
Comment (yoannmoinet) [reply packages/core/src/helpers.ts:230]: Oh, nice catch !

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -90,7 +90,7 @@ jobs:
         key: cache-playwright-binaries-${{ hashFiles('yarn.lock') }}
 
     - uses: actions/upload-artifact@v4
-      if: ${{ !cancelled() && failure() }}
+      if: ${{ failure() }}
```
Comment (yoannmoinet) [inline .github/workflows/ci.yaml:93]: This wasn't triggering when needed.

File: global.d.ts
Comment (yoannmoinet) [inline global.d.ts:1]: Mostly about documentation, a little bit about typing.

File: package.json [hunk 4]
```diff
@@ -26,6 +26,7 @@
         "loop-published": "yarn workspaces foreach -A --include \"@datadog/*\" --exclude \"@datadog/build-plugins\"",
         "loop": "yarn loop-published -pti",
         "oss": "yarn cli oss -d packages -l mit",
+        "playwright": "yarn workspace @dd/tests playwright",
```
Comment (yoannmoinet) [inline package.json:29]: Just an alias.

File: packages/core/src/helpers.ts [hunk 5]
```diff
@@ -36,9 +39,12 @@ export const formatDuration = (duration: number) => {
...
     const milliseconds = d.getUTCMilliseconds();
-    return `${days ? `${days}d ` : ''}${hours ? `${hours}h ` : ''}${minutes ? `${minutes}m ` : ''}${
-        seconds ? `${seconds}s ` : ''
-    }${milliseconds ? `${milliseconds}ms` : ''}`.trim();
+    const timeString =
+        `${days ? `${days}d ` : ''}${hours ? `${hours}h ` : ''}${minutes ? `${minutes}m ` : ''}${
+            seconds ? `${seconds}s` : ''
+        }`.trim();
+    // Split here so we can show 0ms in case we have a duration of 0.
+    return `${timeString}${!timeString || milliseconds ? ` ${milliseconds}ms` : ''}`.trim();
```
Comment (yoannmoinet) [inline packages/core/src/helpers.ts:47]: It was printing `0ms` all the time.
So I fixed it and added tests for it.

File: packages/core/src/helpers.ts [hunk 6]
```diff
@@ -411,3 +424,145 @@ export const unserializeBuildReport = (report: SerializedBuildReport): BuildRepo
         outputs,
     };
 };
+
```
Comment (yoannmoinet) [inline packages/core/src/helpers.ts:427]: After this, it's just moving helpers to a more central place.

File: packages/plugins/build-report/src/esbuild.ts
Comment (yoannmoinet) [inline packages/plugins/build-report/src/esbuild.ts:1]: There was an issue with external dependencies.
Either ignored, missed or broken.
These changes in the `plugins/build-report` are about supporting external dependencies in all the bundlers.

Also added this to the tests.

File: packages/plugins/build-report/src/helpers.ts [hunk 7]
```diff
@@ -114,7 +101,7 @@ export const cleanName = (context: GlobalContext, filepath: string) => {
             .split('!')
             .pop()!
             // Remove outDir's path.
-            .replace(context.bundler.outDir, '')
+            .replace(getAbsolutePath(context.cwd, context.bundler.outDir), '')
```
Comment (yoannmoinet) [inline packages/plugins/build-report/src/helpers.ts:104]: This is coming from a pretty deep issue with rollup and vite.

Commenting about it here as it's the first place it's visible.

--- 

Rollup doesn't really have a notion of `cwd`, or, it's always using `process.cwd()`.

In our `bundler-plugin` we try to streamline the `cwd` used in the build, so for rollup (and vite) it's a bit more complicated to actually compute, especially impactful when the tests' builds are executed from a place (process.cwd()) and the files are living in a separate place (temporary `workingDir`).

It breaks this computation, which breaks some other things, mostly formatting of outputs.

So now, instead of relying on `process.cwd()`, which is still there as the initial default, and the nearest common directory between outDir, cwd and entry files. 
We first try to find the highest `package.json` file from the `outDir`, and only fallback to the previous process if we could not find it.

It results in a more reliable way of computing a somewhat "accurate" cwd.

File: packages/plugins/build-report/src/xpack.ts
Comment (yoannmoinet) [inline packages/plugins/build-report/src/xpack.ts:1]: Supporting externals.

File: packages/plugins/bundler-report/src/index.ts [hunk 8]
```diff
@@ -2,39 +2,27 @@
...
-
-    for (let i = 0; i < minLength; i++) {
-        // We use the first path as our basis.
-        const component = splitPaths[0][i];
-        if (splitPaths.every((parts) => parts[i] === component)) {
-            commonParts.push(component);
-        } else {
-            break;
-        }
+// Compute the CWD based on a list of directories and the outDir.
```
Comment (yoannmoinet) [inline packages/plugins/bundler-report/src/index.ts:15]: cf: https://github.com/DataDog/build-plugins/pull/145/files#r1991128004

For a more detailed explanation.

File: packages/plugins/bundler-report/src/index.ts [hunk 9]
```diff
@@ -96,12 +88,14 @@ export const getBundlerReportPlugins = (context: GlobalContext): PluginOptions[]
...
+                    const outputOptions = Array.isArray(options.output)
+                        ? options.output
+                        : [options.output];
+                    for (const output of outputOptions) {
+                        handleOutputOptions(output);
+                    }
                 }
             },
-            outputOptions(options) {
-                handleOutputOptions(options);
```
Comment (yoannmoinet) [inline packages/plugins/bundler-report/src/index.ts:103]: This (and above) resulted in a double execution.

It seems that `outputOptions` is just a hook called for EACH output option (it can be an array) passed to the configuration.

Where `options` is the "raw" full option object.

File: packages/plugins/injection/src/esbuild.ts
Comment (yoannmoinet) [inline packages/plugins/injection/src/esbuild.ts:1]: For esbuild we're now writing the stub injection file into a temporary dir.

So we don't have to delete it after the build, and sub-builds won't fail when trying to resolve it without our resolver.

It happens because we mutate the `build.initialOptions.inject` with a "virtual" file that needs to be physically present in the file system because esbuild seems to try to resolve it when validating its options/entries, and only then our loader/resolver can take over.

Some other plugins (out of our scope/ecosystem) would implement sub builds, like for workers, inline conversions, etc... re-using the `build.initialOptions` that we mutated with an `inject` entry that is "virtual", but they would not re-use the `plugins` were our loader/resolver are.

Resulting in a resolution error on the file.

Now, the file always exists, no more errors.

File: packages/plugins/injection/src/index.ts
Comment (yoannmoinet) [inline packages/plugins/injection/src/index.ts:1]: Cleaning, factorising and "only do what's necessary" update.

Some hooks were not universally necessary, `resolveId`, `loadInclude`, `load`.
So I removed them from here, and implemented the necessary steps in each bundler (where necessary).

Also grouped particularities together (for `xpack` for instance).

File: packages/plugins/injection/src/xpack.ts
Comment (yoannmoinet) [inline packages/plugins/injection/src/xpack.ts:1]: Same kind of issue than the `esbuild` counterpart.

We need a stub file in the file system for `xpack`.

But it can't be in a temporary directory like `esbuild` because `xpack` won't resolve it correctly out of its context.

Also, using the `done` hook to delete the file wasn't correct, as it was only executed on success.

BUT, using `shutdown` in async mode with `rspack` would randomly fail.
AND webpack4 doesn't offer this hook, so we fallback to both `done` and `failed`.

Cross bundlers implementations is so cool.

File: packages/tests/jest.config.js
Comment (yoannmoinet) [inline packages/tests/jest.config.js:1]: This was an issue a long time ago.
It started to break some mocks (especially with `memfs` and wasn't actually necessary.

File: packages/tests/playwright.config.ts [hunk 10]
```diff
@@ -28,7 +28,8 @@ export default defineConfig<TestOptions>({
         bundlers: getRequestedBundlers(),
         trace: 'retain-on-failure',
     },
-    timeout: 5_000,
+    globalTimeout: process.env.CI ? 20 * 60 * 1000 : undefined,
+    timeout: 60_000,
```
Comment (yoannmoinet) [inline packages/tests/playwright.config.ts:32]: Giving more space to playwright.

File: packages/tests/src/_jest/fixtures/hard_project/package.json [hunk 11]
```diff
@@ -6,10 +6,5 @@
     "packageManager": "yarn@4.2.1",
     "dependencies": {
         "chalk": "2.3.1"
-    },
-    "devDependencies": {
-        "react": "19.0.0",
-        "react-dom": "19.0.0",
-        "react-router-dom": "6.28.0"
```
Comment (yoannmoinet) [inline packages/tests/src/_jest/fixtures/hard_project/package.json:13]: Not sure why it was there, probably some legacy left overs from the rum plugin's implementation.

File: packages/tests/src/_jest/helpers/runBundlers.ts [hunk 12]
```diff
@@ -223,28 +223,29 @@ export const runBundlers = async (
...
     const bundlersToRun = BUNDLERS.filter(
         (bundler) => !bundlers || bundlers.includes(bundler.name),
     );
 
     const workingDir = await prepareWorkingDir(seed);
 
+    if (NO_CLEANUP) {
+        console.log(`[NO_CLEANUP] Working directory: ${workingDir}`);
+    }
+
```
Comment (yoannmoinet) [inline packages/tests/src/_jest/helpers/runBundlers.ts:237]: Pretty useful log to investigate the directory when you use `NO_CLEANUP=1`.

File: packages/tests/src/unit/factory/helpers.test.ts
Comment (yoannmoinet) [inline packages/tests/src/unit/factory/helpers.test.ts:1]: Update to cover for the rollup/vite's `cwd` compute update.

File: packages/tests/src/unit/plugins/build-report/index.test.ts
Comment (yoannmoinet) [inline packages/tests/src/unit/plugins/build-report/index.test.ts:1]: Update to cover the `external` support.

File: packages/tests/src/unit/tools/src/commands/create-plugin/index.test.ts
Comment (yoannmoinet) [inline packages/tests/src/unit/tools/src/commands/create-plugin/index.test.ts:1]: Same as `resetModules`, doesn't seem necessary anymore.

File: packages/tools/src/commands/prepare-link/index.ts [hunk 13]
```diff
@@ -43,7 +43,11 @@ class PrepareLink extends Command {
                 const pkgJsonPath = path.resolve(ROOT, pkg.location, 'package.json');
                 const pkgJson = require(pkgJsonPath);
                 if (this.revert) {
-                    pkgJson.exports = { '.': './src/index.ts' };
+                    pkgJson.exports = {
+                        './dist/src': './dist/src/index.js',
+                        './dist/src/*': './dist/src/*',
+                        '.': './src/index.ts',
+                    };
```
Comment (yoannmoinet) [inline packages/tools/src/commands/prepare-link/index.ts:50]: Coming from the streamlining update on `./published/**/package.json`.

File: packages/tools/src/helpers.ts
Comment (yoannmoinet) [inline packages/tools/src/helpers.ts:1]: Just a move.

File: packages/tools/src/rollupConfig.mjs
Comment (yoannmoinet) [inline packages/tools/src/rollupConfig.mjs:1]: Better typechecking.

It was actually useless without the `@ts-check` 😆

File: global.d.ts
Comment (yoannmoinet) [inline global.d.ts:1]: Mostly about documentation, a little bit about typing.

File: packages/core/src/helpers.ts (hunk 5 above)
Comment (yoannmoinet) [inline packages/core/src/helpers.ts:47]: It was printing 0ms all the time.
So I fixed it and added tests for it.

File: packages/plugins/build-report/src/esbuild.ts
Comment (yoannmoinet) [inline packages/plugins/build-report/src/esbuild.ts:1]: There was an issue with external dependencies.
Either ignored, missed or broken.
These changes in the plugins/build-report/* are about supporting external dependencies in all the bundlers.

Also added this to the tests.

**Review by elbywan — APPROVED** [review]

**Review by jakub-g — APPROVED** [review]
